### PR TITLE
Bolometry examples

### PR DESCRIPTION
**Pull request by jacklovell**
Add some examples for using the bolometry classes. The examples include:

- Make a simple bolometer camera, either from primitives and relative transforms or from a mesh file and vertex coordinates.
- Calculate the etendue of each detector
- Observe a uniform radiator
- Calculate sensitivity matrices and perform inversions, using both the voxel and ray transfer object techniques.

### CODE REVIEW

**Comment by jacklovell** [comment]
@mattngc this is now ready for review. @vsnever could you please also have a look over the new ray transfer demos in the bolometry section? I want to check that the way I'm extracting the ray transfer information to map back to 2D/3D profiles is the best way (or at least, the recommended way for new users) to do it.

By the way, the unit tests are failing on a completely unrelated test. Looks like the tolerance level for a comparison isn't large enough. Strange that this has only happened recently though.

**Comment by vsnever** [comment]
These demos are very comprehensive and the documentation is superb. Amazing work.

Regarding your question, the `mask` parameter was used to create a ray transfer grid in this demo, so you can use this mask to map the data back to 2D grid. Instead of:
```
for indices, sensitivity in zip(ray_transfer_grid.invert_voxel_map(), summed_sensitivity):
    sensitivity_2d[tuple(np.squeeze(indices).T)] = sensitivity
```
you can do:
```
sensitivity_2d[grid_mask] = summed_sensitivity
```
or:
```
sensitivity_2d[ray_transfer_grid.mask] = summed_sensitivity
```
There is no need to use `invert_voxel_map()` until you have mapped multiple grid cells to the same source with `voxel_map` parameter.

Also, I have some other comments.

1) The call of `plt.show()` in the middle of `observe_radiation_function.py`, `geometry_matrix_with_voxels.py` and `geometry_matrix_with_raytransfer.py` stops execution until the matplotlib window is closed by user. Could you please replace the blocking calls of `plt.show()` with
```
plt.draw()
plt.pause(0.001)
```
and move `plt.show()` to the last line of these demos?

2) The `voxel_data` is not used in `geometry_matrix_with_raytransfer.py`, so this code should be removed:
https://github.com/cherab/core/blob/6f77f53e9ec028312a70181ba0e9b567e4aef89a/demos/observers/bolometry/geometry_matrix_with_raytransfer.py#L234-L240

3) This comment from `geometry_matrix_with_voxels.py` is not relevant to `geometry_matrix_with_raytransfer.py`:
https://github.com/cherab/core/blob/6f77f53e9ec028312a70181ba0e9b567e4aef89a/demos/observers/bolometry/geometry_matrix_with_raytransfer.py#L198-L202

**Comment by vsnever** [comment]
These lines can be removed:
https://github.com/cherab/core/blob/6f77f53e9ec028312a70181ba0e9b567e4aef89a/demos/observers/bolometry/geometry_matrix_with_raytransfer.py#L183
https://github.com/cherab/core/blob/6f77f53e9ec028312a70181ba0e9b567e4aef89a/demos/observers/bolometry/geometry_matrix_with_voxels.py#L183

Also, `all_cell_vertices.take(1, -1).min()` is not zero, so instead of this:
https://github.com/cherab/core/blob/6f77f53e9ec028312a70181ba0e9b567e4aef89a/demos/observers/bolometry/geometry_matrix_with_raytransfer.py#L242-L247
should be this:
```
ray_transfer_grid = RayTransferCylinder(
    radius_outer=all_cell_vertices.take(0, -1).max(),
    radius_inner=all_cell_vertices.take(0, -1).min(),
    height=all_cell_vertices.take(1, -1).max() - all_cell_vertices.take(1, -1).min(),
    n_radius=nx, n_height=ny, mask=grid_mask, n_polar=0,
    transform=translate(0, 0, all_cell_vertices.take(1, -1).min())
)
```

**Comment by vsnever** [comment]
I found a simpler and slightly faster way to do this:
https://github.com/cherab/core/blob/6f77f53e9ec028312a70181ba0e9b567e4aef89a/demos/observers/bolometry/geometry_matrix_with_raytransfer.py#L167-L328
with 63 lines instead of 161:
```python
########################################################################
# Produce a voxel grid
########################################################################
print("Producing the voxel grid...")
# We'll use a grid of rectangular voxels here, all of which are the same
# size. Neither the shape nor the uniform size are required for using the
# voxels, but it makes this example a bit simpler.

# Define the centres of each voxel, as an (nx, ny, 2) array
nx = 40
ny = 60
cell_r, cell_dx = np.linspace(1, 3, nx, retstep=True)
cell_z, cell_dy = np.linspace(0, 3, ny, retstep=True)
cell_r_grid, cell_z_grid = np.broadcast_arrays(cell_r[:, None], cell_z[None, :])
cell_centres = np.stack((cell_r_grid, cell_z_grid), axis=-1)  # (nx, ny, 2) array

# Define the positions of the vertices of the voxels
cell_vertices_r = np.linspace(cell_r[0] - 0.5 * cell_dx, cell_r[-1] + 0.5 * cell_dx, nx + 1)
cell_vertices_z = np.linspace(cell_z[0] - 0.5 * cell_dy, cell_z[-1] + 0.5 * cell_dy, ny + 1)

# Build a mask, only including cells within the wall
radius_squared = ((cell_vertices_r[:, None] - PLASMA_AXIS.x)**2 + (cell_vertices_z[None, :] - PLASMA_AXIS.y)**2)
vertex_mask = radius_squared <= (WALL_LCFS_OFFSET + LCFS_RADIUS)**2
# Cell is included if at least one vertex is within the wall
grid_mask = vertex_mask[1:, :-1] + vertex_mask[:-1, :-1] + vertex_mask[1:, 1:] + vertex_mask[:-1, 1:]

num_cells = grid_mask.sum()

ray_transfer_grid = RayTransferCylinder(
    radius_outer=cell_vertices_r[-1],
    radius_inner=cell_vertices_r[0],
    height=cell_vertices_z[-1] - cell_vertices_z[0],
    n_radius=nx, n_height=ny, mask=grid_mask, n_polar=0,
    transform=translate(0, 0, cell_vertices_z[0])
)

########################################################################
# Produce a regularisation operator for inversions
########################################################################
# We'll use simple isotropic smoothing here, in which case an ND second
# derivative operator (the laplacian operator) is appropriate
voxel_map_with_borders = - np.ones((nx + 2, ny + 2), dtype=int)
voxel_map_with_borders[1:-1, 1:-1] = ray_transfer_grid.voxel_map
inverted_voxel_map = ray_transfer_grid.invert_voxel_map()
grid_laplacian = np.zeros((num_cells, num_cells))


for ith_cell in range(num_cells):

    # get the 2D mesh coordinates of this cell
    ix, iy = inverted_voxel_map[ith_cell]
    # we didn't map multiple cells into the same light source,
    # so ix and iy are single-element arrays
    ix = ix[0]
    iy = iy[0]

    neighbours_2d = ([ix, ix    , ix    , ix + 1, ix + 1, ix + 2, ix + 2, ix + 2],
                     [iy, iy + 1, iy + 2, iy    , iy + 2, iy    , iy + 1, iy + 2])

    neighbours_1d = voxel_map_with_borders[neighbours_2d]
    neighbours_1d = neighbours_1d[neighbours_1d > -1]

    grid_laplacian[ith_cell, neighbours_1d] = -1
    grid_laplacian[ith_cell, ith_cell] = neighbours_1d.size
```
However, the original code contains less numpy tricks and may be clearer for beginners.

**Comment by jacklovell** [comment]
Thanks for the feedback Vlad, this is really helpful. Yes, there are some redundant bits of code which came from copy-pasting the voxel example: I'll remove those. I think it's fine to use your vectorised method for producing the voxel mask and Laplacian, as the other method is already demonstrated in the voxel example. I'll make sure it's clear that users can use either method.

**Comment by mattngc** [comment]
Hey @jacklovell, sorry for the delays with looking at this. Just coming back from holidays and been super busy with work stuff. Looks like @vsnever gave you a very comprehensive review.

There is a lot to look at here so might take me a day or two to go through it properly. But we can't merge it anyway until #200 is resolved (hopefully later today).

**Comment by jacklovell** [comment]
@mattngc no worries. I'll need to update the ray transfer demos anyway now that #198 is merged.

**Comment by mattngc** [comment]
Also, you must obviously check that the build is passing after updating the branch...

File: docs/source/demonstrations/bolometry/camera_from_primitives.rst [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+In this demonstration we build a simple bolometer camera using geometric primitives from
+raysect. The camera itself is a rectangular box, with a rectangular aperture forming the
+slit. There are 4 bolometer foils inside the box. The slit and foil positions and
+orientations are defined relative to the camera. Defining a bolometer system in this way
+is common when working from design drawings.
+
+We use raysect primitives (:class:`Box` and CSG operations) to describe the geometry of
+the system. A camera geometry should always be present, else there is the risk that
+stray light can reach the modelled bolometer foils when it would be blocked by the
```
[the remark this answers is not in the document]
Comment (jacklovell) [reply docs/source/demonstrations/bolometry/camera_from_primitives.rst]: I tried adding the `..Warning::`, but couldn't do it in such a way that the warning box didn't interrupt the narrative. I've settled instead for emphasising the point with bold font so it at least stands out more on the page, and users are less likely to skip over it.

**Pull request merged**